src/coreclr/jit/optimizer.cpp

```diff
@@ -4662,267 +5010,311 @@ PhaseStatus Compiler::optUnrollLoops()
 {
 #ifndef TARGET_ARM64
     return false;
-#else
+#endif
 
     // Don't optimise the block if it is inside a loop
     // When inside a loop, branches are quicker than selects.
     // Detect via the block weight as that will be high when inside a loop.
-    if ((block->getBBWeight(this) > BB_UNITY_WEIGHT) && !compStressCompile(STRESS_IF_CONVERSION_INNER_LOOPS, 25))
+    if ((m_startBlock->getBBWeight(m_comp) > BB_UNITY_WEIGHT) &&
+        !m_comp->compStressCompile(m_comp->STRESS_IF_CONVERSION_INNER_LOOPS, 25))
     {
         return false;
     }
 
     // Does the block end by branching via a JTRUE after a compare?
-    if (block->bbJumpKind != BBJ_COND || block->NumSucc() != 2)
+    if (m_startBlock->bbJumpKind != BBJ_COND || m_startBlock->NumSucc() != 2)
     {
         return false;
     }
 
     // Verify the test block ends with a condition that we can manipulate.
-    GenTree* last = block->lastStmt()->GetRootNode();
+    GenTree* last = m_startBlock->lastStmt()->GetRootNode();
     noway_assert(last->OperIs(GT_JTRUE));
     GenTree* cond = last->gtGetOp1();
     if (!cond->OperIsCompare())
     {
         return false;
     }
-
-    // Block where the flows merge.
-    BasicBlock* finalBlock = block->bbNext;
-    // The node, statement and block of the assignment.
-    GenTree*    asgNode  = nullptr;
-    Statement*  asgStmt  = nullptr;
-    BasicBlock* asgBlock = nullptr;
-
-    // Check the block is followed by a block or chain of blocks that only contain NOPs and
-    // a single ASG statement. The destination of the final block must point to the same as the
-    // true path of the JTRUE block.
-    bool foundMiddle = false;
-    while (!foundMiddle)
+    // Check the condition or inverted condition is usable. Prefer inverted.
+    bool         invertCond = true;
+    GenCondition genCond    = GenCondition::FromRelop(cond);
+    if (!GenCondition::IsPreferredRelop(GenCondition::Reverse(genCond)))
     {
-        BasicBlock* middleBlock = finalBlock;
-        noway_assert(middleBlock != nullptr);
-
-        // middleBlock should have a single successor.
-        finalBlock = middleBlock->GetUniqueSucc();
-        if (finalBlock == nullptr)
+        invertCond = false;
+        if (!GenCondition::IsPreferredRelop(genCond))
         {
             return false;
         }
+    }
 
-        if (finalBlock == block->bbJumpDest)
-        {
-            // This is our final middle block.
-            foundMiddle = true;
-        }
-
-        // Check that we have linear flow and are still in the same EH region
+    // Look for valid flow of Then and Else blocks.
+    IfConvertFindFlow();
+    if (!m_flowFound)
+    {
+        return false;
+    }
 
-        if (middleBlock->GetUniquePred(this) == nullptr)
+    // Check the Then and Else blocks have a single assignment each.
+    if (!IfConvertCheckStmts(m_startBlock->bbNext, &m_thenOperation))
+    {
+        return false;
+    }
+    if (m_doElseConversion)
```

Comment:
I guess the main thing that complicates this is that we try to allow arbitrary number of blocks containing NOPs in the then/else branches. Am I right that this is the main reason `IfConvertFindFlow` exists and needs to loop? Does this get us many cases? It seems it would simplify things a lot if we only allowed simple flow like

```
         StartBlock
          /      \
       Then       |
          \      /
          FinalBlock
```
and

```
         StartBlock
          /      \
       Then     Else
          \      /
          FinalBlock
```

Reply:
>Am I right that this is the main reason IfConvertFindFlow exists and needs to loop?

Yes

>  Does this get us many cases? 

Looking at spmi asm.libraries.pmi.Linux.arm64.checked it gets us a few.

I edited my code so that `m_checkLimit = JitConfig.JitDoIfConversion();` (the number of blocks allowed in the true and else paths). Log file number is that value.

```
superpmi.1.log:[17:06:35] Total bytes of delta: -50636 (-0.08% of base)
superpmi.2.log:[17:11:55] Total bytes of delta: -51284 (-0.08% of base)
superpmi.3.log:[17:23:42] Total bytes of delta: -51308 (-0.08% of base)
superpmi.4.log:[17:28:45] Total bytes of delta: -51308 (-0.08% of base)
```

Examining a little, it does seem to be a mixture of empty blocks and blocks with NOPs being hit.

> It seems it would simplify things a lot if we only allowed simple flow

Agreed :)

Is there a function to strip all the empty blocks and NOP blocks from the graph? If so, that'd be an alternative.

However, if in the future we want to capture multiple statements, eg: `if (cond) { x=1; y=2; } else { y=3; z=4; a=2; }`, then this looping logic would be required.